ml-agents/mlagents/trainers/training_status.py

```diff
@@ -0,0 +1,114 @@
+
+    def to_dict(self) -> Dict[str, str]:
+        return cattr.unstructure(self)
+
+    @staticmethod
+    def from_dict(import_dict: Dict[str, str]) -> "StatusMetaData":
+        return cattr.structure(import_dict, StatusMetaData)
+
+    def check_compatibility(self, other: "StatusMetaData") -> None:
+        """
+        Check compatibility with a loaded StatsMetaData and warn the user
+        if versions mismatch. This is used for resuming from old checkpoints.
+        """
+        # This should cover all stats version mismatches as well.
+        if self.mlagents_version != other.mlagents_version:
+            logger.warning(
+                "Checkpoint was loaded from a different version of ML-Agents. Some things may not resume properly."
+            )
+        if self.tensorflow_version != other.tensorflow_version:
+            logger.warning(
+                "Tensorflow checkpoint was saved with a different version of Tensorflow. Model may not resume properly."
+            )
+
+
+class GlobalTrainingStatus:
+    saved_state: Dict[str, Dict[str, Any]] = defaultdict(lambda: {})
+
+    def __init__(self, category: str):
+        """
+        Generic GlobalTrainingStatus writer. A category is the broadest type of storage (would
+        correspond the run name and trainer name, e.g. 3DBalltest_3DBall. A key is the
+        type of status needed to be saved (e.g. Lesson Number). Finally the Value is the float value
+        attached to this stat.
+        """
+        self.category: str = category
+
+    @staticmethod
+    def load_state(path: str) -> None:
+        """
+        Load a JSON file that contains saved state.
+        :param path: Path to the JSON file containing the state.
+        """
+        try:
+            with open(path, "r") as f:
+                loaded_dict = json.load(f)
+            # Compare the metadata
+            _metadata = loaded_dict[StatusType.STATS_METADATA.value]
+            StatusMetaData.from_dict(_metadata).check_compatibility(StatusMetaData())
+            # Update saved state.
+            GlobalTrainingStatus.saved_state.update(loaded_dict)
+        except FileNotFoundError:
+            logger.warning(
+                "Training status file not found. Not all functions will resume properly."
+            )
+        except KeyError:
+            raise TrainerError(
+                "Metadata not found, resuming from an incompatible version of ML-Agents."
+            )
+
+    @staticmethod
+    def save_state(path: str) -> None:
+        """
+        Save a JSON file that contains saved state.
+        :param path: Path to the JSON file containing the state.
+        """
+        GlobalTrainingStatus.saved_state[
+            StatusType.STATS_METADATA.value
+        ] = StatusMetaData().to_dict()
+        with open(path, "w") as f:
+            json.dump(GlobalTrainingStatus.saved_state, f, indent=4)
+
+    def store_parameter_state(self, key: StatusType, value: Any) -> None:
```

Comment:
nit: `set_parameter_state()` and `get_paramater_state()`? I think `restore_parameter_state` is a bad name since "restore" makes it sound like it's doing loading.

Reply:
Changed.